### PR TITLE
Combine autograd function and method tests, preferring method tests

### PR DESCRIPTION
**Pull request by gchanan**
The goal of this PR is to leave a single procedure for specifying common autograd tests; prior to this there are both Function and method tests, which is error prone because:
1) tests need to be specified in two places
2) each way tests something different, e.g. method tests didn't test autograd, so in some cases weren't specified properly (e.g. division by 0)

Here, we prefer method tests because Functions are really an implementation detail, and relatedly we are working to remove the python overhead by moving most of the Function code to C++.

The mechanical part of this PR is basically:
1) Moving the test logic that is not duplicated from the Function tests to the method tests
2) Moving or de-duplicating the Function tests to method test equivalents
3) Small cleanups:
a) Making tensor and variable equivalent (i.e. adding __rpow__ to tensor, as it exists on Variable)
b) Moving Concat function tests to individual test cases because it doesn't have a Variable method and doesn't fit clearly in anywhere else.
c) cleaning up some method tests that failed the new checks, as mentioned above.

### CODE REVIEW

**Comment by gchanan** [comment]
cc @colesbury

File: test/test_autograd.py [hunk 1]
```diff
@@ -1485,6 +1485,36 @@ def backward(ctx, grad_output):
         out.sum().backward()
         self.assertEqual(x.grad.data, y_data)
 
+    def test_cat(self):
```
Comment (ezyang) [inline test/test_autograd.py]: I'm impressed that this is the only function that doesn't have a method, that didn't already have a standalone test :)

File: torch/tensor.py [hunk 2]
```diff
@@ -320,6 +320,9 @@ def __matmul__(self, other):
     def __pow__(self, other):
         return self.pow(other)
 
+    def __rpow__(self, other):
```
Comment (ezyang) [inline torch/tensor.py]: Nice bonus feature ;)

**Comment by ezyang** [comment]
Very nice!

**Pull request merged**